parachain/pallets/omni-account/src/lib.rs

```diff
@@ -0,0 +1,218 @@
+pub struct MemberAccount {
+	pub id: MemberIdentity,
+	pub hash: H256,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The event type of this pallet.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// The origin which can manage the pallet.
+		type TEECallOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+		/// The maximum number of identities an id graph can have.
+		#[pallet::constant]
+		type MaxIDGraphLength: Get<u32>;
+	}
+	pub type IDGraphMembers<T> = BoundedVec<MemberAccount, <T as Config>::MaxIDGraphLength>;
+
+	#[pallet::storage]
+	pub type LinkedIdentityHashes<T: Config> =
+		StorageMap<Hasher = Blake2_128Concat, Key = H256, Value = ()>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn id_graphs)]
+	pub type IDGraphs<T: Config> =
+		StorageMap<Hasher = Blake2_128Concat, Key = Identity, Value = IDGraphMembers<T>>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn id_graph_hashes)]
+	pub type IDGraphHashes<T: Config> =
+		StorageMap<Hasher = Blake2_128Concat, Key = Identity, Value = H256>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Identity linked
+		IdentityLinked { who: Identity, identity: H256 },
+		/// Identity remove
+		IdentityRemoved { who: Identity, identity_hashes: Vec<H256> },
+		/// Identity made public
+		IdentityMadePublic { who: Identity, identity_hash: H256 },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Identity is already linked
+		IdentityAlreadyLinked,
+		/// IDGraph len limit reached
+		IDGraphLenLimitReached,
+		/// Identity not found
+		IdentityNotFound,
+		/// Invalid identity
+		PrimeIdentityInvalid,
+		/// Prime identity not found
+		PrimeIdentityNotFound,
+		/// Identity is private
+		IdentityIsPrivate,
+		/// Identities empty
+		IdentitiesEmpty,
+		/// IDGraph hash does not match
+		IDGraphHashMismatch,
+		/// Missing IDGraph hash
+		IDGraphHashMissing,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight((195_000_000, DispatchClass::Normal))]
+		pub fn link_identity(
+			origin: OriginFor<T>,
+			who: Identity,
+			identity_hash: H256,
+			identity: MemberIdentity,
+			maybe_id_graph_hash: Option<H256>,
+		) -> DispatchResult {
+			let _ = T::TEECallOrigin::ensure_origin(origin)?;
+			ensure!(
+				!LinkedIdentityHashes::<T>::contains_key(identity_hash),
+				Error::<T>::IdentityAlreadyLinked
+			);
+			let mut id_graph_members = match IDGraphs::<T>::get(&who) {
+				Some(id_graph_members) => {
+					let current_id_graph_hash =
+						IDGraphHashes::<T>::get(&who).ok_or(Error::<T>::PrimeIdentityNotFound)?;
```

Comment:
Why don't you use `IDGraphHashMissing` ?

Reply:
I guess it could be ok, but `IDGraphHashMissing` has a different meaning. It means that the the user wanted to update an existing IDGraph but it didn't provide the IDGraph hash (maybe_id_graph_hash param). In theory this error should never happen because if the IDGraph for `who` exists, the IDGraphHash should also exist. So, this would be a logic error